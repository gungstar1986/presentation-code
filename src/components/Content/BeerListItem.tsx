import React, { forwardRef } from 'react';
import { Link } from 'react-router-dom';

import { CONTENT } from 'constants/routes';
import { ItemListType } from './interfaces';

export const BeerListItem: ItemListType = forwardRef(({ id, className, handleShowContentItem, image, name }, ref) => {
  return (
    <div className={className} onClick={handleShowContentItem.bind(null, id)} ref={ref}>
      <Link className="content-link" to={`${CONTENT}/${id}`}>
        {name}
      </Link>
      <div className="link-img-container">
        <img className="content-link-img" src={image} alt={name} />
      </div>
    </div>
  );
});
