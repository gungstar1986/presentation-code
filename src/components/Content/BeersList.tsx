import React from 'react';
import { observer } from 'mobx-react-lite';

import { useBeerStore } from 'hooks/storesHooks/useBeerStore';
import { useRedirectToFirstItem } from 'hooks/customHooks/useRedirectToFirstItem';

import { LoadButtons } from 'components/common/LoadButtons';
import { ScrollContainer } from 'components/common/ScrollContainer';
import { MAX_CONTENT_ITEMS, MIN_CONTENT_ITEMS } from 'store/beer/constants';
import { NEXT_BUTTON, PREV_BUTTON } from 'components/common/LoadButtons/constants';
import { ContentListWithRefItem } from 'components/common/ContentListWithRefItem';

export const BeersList: React.FC = observer(() => {
  const { beerList, fetchBeerById, beerListItem, fetchBeerList, fetchPrevItemsList } = useBeerStore();

  useRedirectToFirstItem(beerList);

  const handleFetchBeerItem = async (id: number) => {
    if (beerListItem.id === id) return;
    await fetchBeerById(id);
  };

  const handleLoadNextContent = async () => {
    await fetchBeerList();
  };

  const handleLoadPrevContent = async () => {
    await fetchPrevItemsList();
  };

  const firstItem = MIN_CONTENT_ITEMS < beerList[0].id;
  const lastItem = MAX_CONTENT_ITEMS > beerList[beerList.length - 1].id;

  return (
    <div className="content-item">
      <ScrollContainer>
        <LoadButtons loadFn={handleLoadPrevContent} title={PREV_BUTTON} isRender={firstItem} />
        <ContentListWithRefItem loadFn={handleFetchBeerItem} content={beerList} contentItem={beerListItem} />
        <LoadButtons loadFn={handleLoadNextContent} title={NEXT_BUTTON} isRender={lastItem} />
      </ScrollContainer>
    </div>
  );
});
