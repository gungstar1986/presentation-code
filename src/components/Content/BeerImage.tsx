import React from 'react';

import { BeerImgInterface } from 'components/Content/interfaces';

export const BeerImage: React.FC<BeerImgInterface> = ({ image, name }) => {
  return (
    <div className="content-item center">
      <div className="media">
        <img className="media-content" src={image} alt={name} />
      </div>
    </div>
  );
};
