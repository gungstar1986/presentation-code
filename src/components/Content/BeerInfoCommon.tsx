import React, { useEffect } from 'react';
import { observer } from 'mobx-react-lite';
import { useParams } from 'react-router-dom';

import { useBeerStore } from 'hooks/storesHooks/useBeerStore';

import { pageCounter } from 'helpers/pageCounter';
import { BeerImage } from 'components/Content/BeerImage';
import { BeerDescription } from 'components/Content/BeerDescription';

export const BeerInfoCommon: React.FC = observer(() => {
  const {
    beerListItem,
    fetchBeerById,
    paginate: { per_page },
    fetchContentByCustomPaginate,
  } = useBeerStore();

  const { beerId } = useParams();

  useEffect(() => {
    const isCustomPaginate = Number(beerId) > per_page && !beerListItem;

    if (isCustomPaginate) {
      const paginate = pageCounter({ beerId: Number(beerId), perPage: per_page });
      fetchContentByCustomPaginate(paginate);
    }

    if (!beerListItem) fetchBeerById(beerId);
  }, []);

  if (!beerListItem) return null;

  const { image_url, name } = beerListItem;

  return (
    <>
      <BeerImage name={name} image={image_url} />
      <BeerDescription descript={beerListItem} />
    </>
  );
});
