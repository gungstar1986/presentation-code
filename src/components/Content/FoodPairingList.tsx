import React from 'react';
import { v4 as uuid } from 'uuid';

import { FoodPairingProps } from 'components/Content/interfaces';

export const FoodPairingList: React.FC<FoodPairingProps> = ({ food_pairing }) => {
  const foodList = food_pairing.map((food, index) => {
    return (
      <span key={uuid()}>
        {index}: {food}
      </span>
    );
  });

  return <>{foodList}</>;
};
