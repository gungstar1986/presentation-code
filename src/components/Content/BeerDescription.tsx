import React from 'react';

import { HopList } from 'components/Content/HopList';
import { BeerDescript } from 'components/Content/interfaces';
import { ScrollContainer } from 'components/common/ScrollContainer';
import { FoodPairingList } from 'components/Content/FoodPairingList';

export const BeerDescription: React.FC<BeerDescript> = ({ descript }) => {
  const { volume, description, first_brewed, food_pairing, ingredients, boil_volume, brewers_tips } = descript;

  const { hops } = ingredients;

  return (
    <div className="content-item right">
      <ScrollContainer>
        <div className="content-link-container active">{brewers_tips}</div>
        <div className="content-link-container active">{description}</div>
        <div className="content-link-container active">{first_brewed}</div>
        <div className="content-link-container active">
          <div className="list">
            <FoodPairingList food_pairing={food_pairing} />
          </div>
        </div>
        <div className="content-link-container active">
          <div className="list">
            <HopList hops={hops} />
          </div>
        </div>
        <div className="content-link-container active">
          {volume.unit} {volume.value}
        </div>
        <div className="content-link-container active">
          {boil_volume.unit} {boil_volume.value}
        </div>
      </ScrollContainer>
    </div>
  );
};
