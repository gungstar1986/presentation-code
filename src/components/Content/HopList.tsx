import React from 'react';
import { v4 as uuid } from 'uuid';

import { HopListProps } from 'components/Content/interfaces';

export const HopList: React.FC<HopListProps> = ({ hops }) => {
  const hopList = hops.map((hop) => {
    return (
      <div key={uuid()}>
        <div>{hop.name}</div>
        <div>
          {hop.amount.unit}: {hop.amount.value}
        </div>
        <div>{hop.attribute}</div>
      </div>
    );
  });

  return <>{hopList}</>;
};
