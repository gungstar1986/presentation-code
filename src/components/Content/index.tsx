import React, { useEffect } from 'react';
import { Route } from 'react-router-dom';
import { observer } from 'mobx-react-lite';

import { useBeerStore } from 'hooks/storesHooks/useBeerStore';

import { BeersList } from 'components/Content/BeersList';
import { CONTENT, BEER_FROM_COLLECTION } from 'constants/routes';
import { BeerInfoCommon } from 'components/Content/BeerInfoCommon';

export const ContentMainComponent: React.FC = observer(() => {
  const { fetchBeerList, beerList } = useBeerStore();

  useEffect(() => {
    if (!beerList) fetchBeerList();
  }, []);

  if (!beerList) return null;

  return (
    <div className="main-content-container">
      <Route path={CONTENT} component={BeersList} />
      <Route exact path={BEER_FROM_COLLECTION} component={BeerInfoCommon} />
    </div>
  );
});
