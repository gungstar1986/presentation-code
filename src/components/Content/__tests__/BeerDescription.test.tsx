import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { mockBeerData } from '_mock/mock';
import { BeerDescription } from 'components/Content/BeerDescription';

describe('BeerDescription component', () => {
  it('should render BeerDescription component', () => {
    const wrapper = render(
      <BrowserRouter>
        <BeerDescription descript={mockBeerData} />
      </BrowserRouter>
    );

    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
