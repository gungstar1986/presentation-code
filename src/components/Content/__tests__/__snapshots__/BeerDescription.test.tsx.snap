// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerDescription component should render BeerDescription component 1`] = `
<body>
  <div>
    <div
      class="content-item right"
    >
      <div
        class="scrollbar-container ps"
      >
        <div
          class="content-link-container active"
        >
          While it may surprise you, this version of Punk IPA isn't dry hopped but still packs a punch! To make the best of the aroma hops make sure they are fully submerged and add them just before knock out for an intense hop hit.
        </div>
        <div
          class="content-link-container active"
        >
          Our flagship beer that kick started the craft beer revolution. This is James and Martin's original take on an American IPA, subverted with punchy New Zealand hops. Layered with new world hops to create an all-out riot of grapefruit, pineapple and lychee before a spiky, mouth-puckering bitter finish.
        </div>
        <div
          class="content-link-container active"
        >
          04/2007
        </div>
        <div
          class="content-link-container active"
        >
          <div
            class="list"
          >
            <span>
              0
              : 
              Spicy carne asada with a pico de gallo sauce
            </span>
            <span>
              1
              : 
              Shredded chicken tacos with a mango chilli lime salsa
            </span>
            <span>
              2
              : 
              Cheesecake with a passion fruit swirl sauce
            </span>
          </div>
        </div>
        <div
          class="content-link-container active"
        >
          <div
            class="list"
          >
            <div>
              <div>
                Ahtanum
              </div>
              <div>
                grams
                : 
                17.5
              </div>
              <div>
                bitter
              </div>
            </div>
            <div>
              <div>
                Chinook
              </div>
              <div>
                grams
                : 
                15
              </div>
              <div>
                bitter
              </div>
            </div>
          </div>
        </div>
        <div
          class="content-link-container active"
        >
          liters
           
          20
        </div>
        <div
          class="content-link-container active"
        >
          liters
           
          25
        </div>
        <div
          class="ps__rail-x"
          style="left: 0px; top: 0px;"
        >
          <div
            class="ps__thumb-x"
            style="left: 0px; width: 0px;"
            tabindex="0"
          />
        </div>
        <div
          class="ps__rail-y"
          style="top: 0px; left: 0px;"
        >
          <div
            class="ps__thumb-y"
            style="top: 0px; height: 0px;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
